cpp/tensorrt_llm/batch_manager/kvCacheManager.cpp

```diff
@@ -354,132 +401,749 @@ std::optional<std::chrono::milliseconds> KVCacheBlock::getDurationMs() const
+            newNode->setPrevNode(searchRoot);
+            searchRoot->addNextNode(blockKey, newNode);
+            matches.emplace_back(
+                std::make_tuple(isFull, static_cast<SizeType32>(blockKey.uniqueTokens.size()), newNode));
+            searchRoot = newNode;
+        }
+        else if (matches.empty())
+        {
+            // No match
+            // Stop search here
+            searchRoot = nullptr;
+        }
+        else
+        {
+            // Stop search if first node is a partial match
+            // TODO: This may not be correct, findMatchingNodes can return both a full match and several partial matches
+            // at once.
+            auto const& [partialMatch, _, matchingNode] = matches[0];
+            searchRoot = partialMatch ? nullptr : matchingNode;
+        }
+        results.emplace_back(std::move(matches));
+    }
+    return results;
+}
+
+// Return BlockKey of first new context block for all window sizes. If no new context block was found for a window size,
+// return value will not have an entry for that window size.
+std::unordered_map<SizeType32, BlockKey> KVCachePromptLookup::findNewContextBlock(LlmRequest const& llmRequest,
+    SizeType32 inputLength, bool allowPartiallyFilledBlock, std::vector<SizeType32> const& windowSizes) const
+{
+    auto blockKeys = getBlockKeys(llmRequest, inputLength, allowPartiallyFilledBlock);
+
+    // New context block is the block key of the first block that isn't found in search structure.
+    std::unordered_set<SizeType32> stillLooking;
+    for (auto const windowSize : windowSizes)
+    {
+        stillLooking.insert(windowSize);
+    }
+    std::unordered_map<SizeType32, BlockKey> results;
+    auto searchRoot = mRoot;
+    BlockKey prevBlockKey;
+    for (auto const& blockKey : blockKeys)
+    {
+        auto matches = searchRoot != nullptr ? searchRoot->findMatchingNodes(blockKey, false, true) : LookupResult();
+        [[maybe_unused]] auto const& [dummy1, dummy2, matchingNode]
+            = matches.empty() ? std::make_tuple(false, 0, nullptr) : matches[0];
+        for (auto const windowSize : windowSizes)
+        {
+            if (stillLooking.count(windowSize)
+                && (matchingNode == nullptr || matchingNode->getBlock(windowSize) == nullptr))
+            {
+                // Found new context block for current window size
+                results[windowSize] = blockKey;
+                stillLooking.erase(windowSize);
+            }
+        }
+        searchRoot = matchingNode;
+        if (stillLooking.size() == 0)
+        {
+            // Not looking for any more new context block
+            break;
+        }
+    }
+
+    return results;
 }
 
-void KVCacheBlock::setExpirationTime(std::optional<std::chrono::steady_clock::time_point::duration> expirationTime)
+KVCachePromptLookupNode::KVCachePromptLookupNode(BlockKey const& blockKey, bool isFull, bool isRoot)
+    : mBlockKey{blockKey}
```

Comment:
Pass by value and use std::move (clang-tidy modernize-pass-by-value)
```suggestion
KVCachePromptLookupNode::KVCachePromptLookupNode(BlockKey blockKey, bool isFull, bool isRoot)
    : mBlockKey{std::move(blockKey)}
```